documentation/High-level-overview.md

```diff
@@ -0,0 +1,150 @@
+
+The first step of evaluation is to load the project file and the XML data it contains. There are multiple passes within the same file to collect data, some of those to specifically define project properties and imports that are necessary for other tasks. At this time, the restore target has run already, so all imports are files on disk and are processed as paths by the engine. Another characteristic of imports is that they are brough within the project logic, so other projects can refence the same import logic instead of having a copy of the same data. Data loaded within the evaluation are not used until execution stage. This means that data can be added and modified during evaluation. 
+
+The evaluation stage should not have any side effect on disk, no new or deleted files. Two exceptions for this are:
+ - SDK resolution
+ - NuGet SDK, which might add packages to the disk
+
+### imports
+Complex projects generally include imports of many different types. In MSBuild an import can be a lot of things, a disk path, a property expansion, a known folder, or even environmental variables. There are also some main imports that come with the execution on other platforms, like the Visual Studio or SDK can have import directories that contain wild card imports. However, when it comes to the evaluation phase in MSBuild, imports are all treated like a property plus path expansion, this includes imported NuGet packages.
+
+In the case of tool imports, MSBuild does not process tool resolution via registry. Instead, it is resolved by looking on adjacent folder to the current running version of MSBuild. The folders will be different depending is MSBuild is running from Visual Studio or the .NET SDK.
+
+## Execution operation
+For more detailed information on execution phase visit [Microsoft Learn](https://learn.microsoft.com/en-us/visualstudio/msbuild/build-process-overview#execution-phase).
+
+The execution phase is simply executing the targets defined in the XML by the user or implicitly defined by the SDK or VS. The order of executed targets is defined using a few attributes: `BeforeTargets`, `DependsOnTargets`, and `AfterTargets`. But the final order might change if an earlier target modifies a property of a later target. The full executing order can be [found here](https://learn.microsoft.com/en-us/visualstudio/msbuild/target-build-order#determine-the-target-build-order).
+
+### Task Host
+MSBuild has a tool called Task Host, that allows tasks to run in a different .NET environment than the one used for build execution.
+
+This is an opt-in behavior that can be used for various cases:
+- If a task breaks the build process it can be relegated to the Task Host, so it does not influence the main build.
```

Comment:
Is this something that happens during the execution or it is up to the user to do that explicitly per task?

Reply:
From my understanding it is up to the user to explicitly opt-in the Task Host behaviour per task.